### PR TITLE
Wavefront metrics sink

### PR DESCRIPTION
**Pull request by ezeev**
related: https://github.com/kubernetes/heapster/issues/1361

### CODE REVIEW

**Comment by ezeev** [comment]
Closed and re-opened so the bot picks up signed cncf-cla

**Comment by piosz** [comment]
@ezeev thanks for the contribution. We need someone familiar with the  Wavefront to review it. Do you know such person?

cc @mrcrgl

**Comment by mrcrgl** [comment]
@piosz thank you for mention me. I could have a look at it.

**Comment by piosz** [comment]
@mrcrgl Cool, thanks! Will you be interested in becoming sink owner (see https://github.com/kubernetes/heapster/blob/master/docs/sink-owners.md) once this is merged?

**Review by mrcrgl — COMMENTED** [review]
(removed duplicate)

File: metrics/sinks/wavefront/wavefront.go
Comment (mrcrgl) [inline metrics/sinks/wavefront/wavefront.go]: It may be useful append information to the error like Flag name.

File: metrics/sinks/wavefront/wavefront.go
Comment (mrcrgl) [inline metrics/sinks/wavefront/wavefront.go]: Like above

**Review by mrcrgl — COMMENTED** [review]
Please do not use this for Receiver Names
See: https://github.com/golang/go/wiki/CodeReviewComments#receiver-names

**Comment by mrcrgl** [comment]
@ezeev please, add some tests for your Sink, too.

**Comment by mrcrgl** [comment]
I've finished my review. 

@piosz thank you for asking me being the sink owner. That sounds interesting, lets see how it goes with this PR.

**Comment by ezeev** [comment]
@mrcrgl @piosz - Thank you for getting to this!

I'm an employee at Wavefront so yes I am quite familiar with it. I would like to be the sink owner or at least collab with @piosz closely.

@piosz - thanks for the review! Also - perhaps we can synch together on getting this running in your Wavefront instance?

**Review by mrcrgl — CHANGES_REQUESTED** [review]

File: metrics/sinks/wavefront/wavefront.go
Comment (ezeev) [inline metrics/sinks/wavefront/wavefront.go]: I don't disagree, other than the maps being of different types. We would still need three separate functions because they have different signatures. We could add a single function that accepts a map[string]interface{} then do type assertion. This may be less performant though.

**Comment by ezeev** [comment]
Hi @mrcrgl 

Thanks again for reviewing this. Please take a look at the changes and let me know. Also, I'll send you an email about getting this setup on your Wavefront instance. We're maintaining a fork of Heapster with this sink for users currently.

Thanks,
Evan

**Comment by ezeev** [comment]
Sorry, when I say "fork" - I really meant to say that we're maintaining a container image of heapster packaged w/ this sink (for now).

File: metrics/sinks/wavefront/wavefront.go
Comment (mrcrgl) [inline metrics/sinks/wavefront/wavefront.go]: I see your point

**Comment by mrcrgl** [comment]
Hi @ezeev 

Thank you for applying my notes. It's fine for me if you would be the sink owner, or we both together. 
We're actually using your container image of Heapster in our test-environment and looking forward to use it in production as well.

Best,
Marc

**Comment by ezeev** [comment]
@mrcrgl @piosz is this ok to merge?

**Comment by mrcrgl** [comment]
Hi @ezeev,

It'd be great to have some basic tests, what do you think?

Best,
Marc

**Comment by piosz** [comment]
It needs LGTM from @mrcrgl to be merged.

**Comment by DirectXMan12** [comment]
ping @mrcrgl

**Comment by mrcrgl** [comment]
Hi @ezeev, the looks fine to me but I'd like to have some unit tests added. 
In my opinion it is easier and safer to maintain the code afterwards. 

/cc @DirectXMan12

**Comment by DirectXMan12** [comment]
@mrcrgl agreed about the unit tests

@ezeev alright, please add some unit tests, and then I think this has the LGTM stamp after those are done (should be something like "given these metrics in Heapster internal format, do we get the right output metrics in Wavefront format").

Once those tests are in place (and I've had a chance to glance over them), I'll merge this.

**Comment by ezeev** [comment]
Thanks @DirectXMan12 @mrcrgl 

I'm working on the tests now. They will include a test that ensure metrics are in the proper line format for Wavefront.

**Comment by ezeev** [comment]
@DirectXMan12 any idea what caused this travis-ci error? the last commit was a very small change I tested locally. Did something timeout?

**Comment by ezeev** [comment]
@mrcrgl let me know if this LGTY thank you!

**Comment by mrcrgl** [comment]
@ezeev LGTM! Thank you for writing the unit tests. 

I guess this one could be merged, what do you think @piosz?

/cc @DirectXMan12

**Comment by DirectXMan12** [comment]
Generally LGTM.  Please squash the commits together and add a commit updating the state of the Wavefront sink in the `sink-owners.md`.  Thanks!

**Comment by ezeev** [comment]
Hi @DirectXMan12 

I updated the sink status on the sink owners page. I'm having a hell of a time squashing the commits. I wonder if it is because sink-owners.md did not exist when the original PR was submitted. What do you recommend?

This is the input i'm providing when rebasing from within my repo (`git rebase -i`):

```
pick 25c7273 Wavefront metrics sink
pick 85e6654 corrected license
pick 6034d51 moved labels to tags to its own method
pick 8a50311 removed unneeded comment
pick 9a98bbc removed unnecessary if statement in addLabelTags
pick 9b9d96b cleaned up logging
pick e2f215b Changes based on code review suggestions
pick 5f9ff70 Added test mode
pick b2da696 wavefront sink tests
pick 91328b4 resolving gofmt error
pick 3b3d4cf changed metric value test to ParseFloat from ParseInt
pick 4e93964 triggering ci tests
s 8e71c6e updating sink status
```

The output: 

```
error: could not apply 25c7273... Wavefront metrics sink

When you have resolved this problem, run "git rebase --continue".
If you prefer to skip this patch, run "git rebase --skip" instead.
To check out the original branch and stop rebasing, run "git rebase --abort".
Could not apply 25c7273be5b91cee4f005308d94d8e5554052799... Wavefront metrics sink
```

**Comment by ezeev** [comment]
I forced this through. The latest commit is the squashed version.

**Review by mrcrgl — APPROVED** [review]

**Pull request merged**
